p2p/src/authenticated/lookup/actors/peer/actor.rs

```diff
@@ -0,0 +1,237 @@
+            let rate_limits = rate_limits.clone();
+            move |context| async move {
+                // Allow tracker to initialize the peer
+                tracker.connect(peer.clone(), mailbox.clone()).await;
+
+                // Set the initial deadline to now to start pinging immediately
+                let mut deadline = context.current();
+
+                // Enter into the main loop
+                loop {
+                    select! {
+                        _ = context.sleep_until(deadline) => {
+                            // Periodically send a ping to the peer
+                            Self::send(
+                                &mut conn_sender,
+                                &self.sent_messages,
+                                metrics::Message::new_ping(&peer),
+                                types::Message::Ping,
+                            ).await?;
+
+                            // Reset ticker
+                            deadline = context.current() + self.ping_frequency;
+                        },
+                        msg_control = self.control.next() => {
+                            let msg = match msg_control {
+                                Some(msg_control) => msg_control,
+                                None => return Err(Error::PeerDisconnected),
+                            };
+                            match msg {
+                                Message::Kill => {
+                                    return Err(Error::PeerKilled(peer.to_string()))
+                                }
+                            }
+                        },
+                        msg_high = self.high.next() => {
+                            let msg = Self::validate_msg(msg_high, &rate_limits)?;
+                            Self::send(&mut conn_sender, &self.sent_messages, metrics::Message::new_data(&peer, msg.channel), msg.into())
+                                .await?;
+                        },
+                        msg_low = self.low.next() => {
+                            let msg = Self::validate_msg(msg_low, &rate_limits)?;
+                            Self::send(&mut conn_sender, &self.sent_messages, metrics::Message::new_data(&peer, msg.channel), msg.into())
+                                .await?;
+                        }
+                    }
+                }
+            }
+        });
+        let mut receive_handler: Handle<Result<(), Error>> = self
+            .context
+            .with_label("receiver")
+            .spawn(move |context| async move {
+                loop {
+                    let msg = conn_receiver
+                        .receive()
+                        .await
+                        .map_err(Error::ReceiveFailed)?;
+                    let data = match types::Message::decode_cfg(msg, &(..).into()) {
+                        Ok(types::Message::Ping) => {
+                            self.received_messages
```

Comment:
We should rate limit (like we did with bit_vec) the number of pings we'll process.